protocol/kafka_confluent/v2/protocol.go

```diff
@@ -0,0 +1,225 @@
+
+type Protocol struct {
+	kafkaConfigMap *kafka.ConfigMap
+
+	consumer             *kafka.Consumer
+	consumerTopics       []string
+	consumerRebalanceCb  kafka.RebalanceCb     // optional
+	consumerPollTimeout  int                   // optional
+	consumerErrorHandler func(err kafka.Error) //optional
+	consumerMux          sync.Mutex
+
+	producer             *kafka.Producer
+	producerDeliveryChan chan kafka.Event // optional
+	producerDefaultTopic string           // optional
+
+	// receiver
+	incoming chan *kafka.Message
+}
+
+func New(opts ...Option) (*Protocol, error) {
+	p := &Protocol{
+		consumerPollTimeout: 100,
+		incoming:            make(chan *kafka.Message),
+	}
+	if err := p.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if p.kafkaConfigMap != nil {
+		if p.consumerTopics != nil && p.consumer == nil {
+			consumer, err := kafka.NewConsumer(p.kafkaConfigMap)
+			if err != nil {
+				return nil, err
+			}
+			p.consumer = consumer
+		}
+		if p.producerDefaultTopic != "" && p.producer == nil {
+			producer, err := kafka.NewProducer(p.kafkaConfigMap)
+			if err != nil {
+				return nil, err
+			}
+			p.producer = producer
+		}
+		if p.producer == nil && p.consumer == nil {
+			return nil, fmt.Errorf("at least receiver or sender topic must be set")
+		}
+	}
+
+	if p.kafkaConfigMap == nil && p.producer == nil && p.consumer == nil {
+		return nil, fmt.Errorf("at least one of the following to initialize the protocol must be set: config, producer, or consumer")
+	}
+	if p.producer != nil {
+		p.producerDeliveryChan = make(chan kafka.Event)
+	}
+	return p, nil
+}
+
+func (p *Protocol) applyOptions(opts ...Option) error {
+	for _, fn := range opts {
+		if err := fn(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	if p.producer == nil {
+		return fmt.Errorf("producer client must be set")
+	}
+	defer in.Finish(err)
+
+	kafkaMsg := &kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic:     &p.producerDefaultTopic,
+			Partition: kafka.PartitionAny,
+		},
+	}
+
+	if topic := cecontext.TopicFrom(ctx); topic != "" {
+		kafkaMsg.TopicPartition.Topic = &topic
+	}
+
+	if messageKey := MessageKeyFrom(ctx); messageKey != "" {
+		kafkaMsg.Key = []byte(messageKey)
+	}
+
+	err = WriteProducerMessage(ctx, in, kafkaMsg, transformers...)
+	if err != nil {
+		return err
+	}
+
+	err = p.producer.Produce(kafkaMsg, p.producerDeliveryChan)
+	if err != nil {
+		return err
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
+	if m.TopicPartition.Error != nil {
+		return m.TopicPartition.Error
+	}
+	return nil
+}
+
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	if p.consumer == nil {
+		return fmt.Errorf("the consumer client must be set")
+	}
+	if p.consumerTopics == nil {
+		return fmt.Errorf("the consumer topics must be set")
+	}
+
+	p.consumerMux.Lock()
+	defer p.consumerMux.Unlock()
```

Comment:
can lead to deadlocks since you're holding on it for function lifetime. See comment below

Reply:
Whether the mutex is intended to protect the `OpenInbound` can only be invoked once to avoid launching the consumer multiple times. Cause the saram implementation also uses this mutex for the "consumer loop" on the `OpenInbound`.

Reply:
Ah, got it yes.